### ISSUE TITLE
"which pyenv" breaks after adding .zshrc and .zprofile parameters.

### ISSUE THREAD

**Issue by alwayserik** [pyenv/pyenv#2222]
Installed via Homebrew.

Expected output:
```
eg@Eriks-MacBook-Pro ~ % which pyenv
/usr/local/bin/pyenv
```

**Unexpected output after adding .zshrc and .zprofile modified:**
```
eg@Eriks-MacBook-Pro ~ % which pyenv
pyenv () {
	local command
	command="${1:-}" 
	if [ "$#" -gt 0 ]
	then
		shift
	fi
	case "$command" in
		(rehash | shell) eval "$(pyenv "sh-$command" "$@")" ;;
		(*) command pyenv "$command" "$@" ;;
	esac
}
```
pyenv works appropriately otherwise, PATH variable is modified appropriately for shims and so on etc.

### Description
- [ ] Platform information (e.g. Ubuntu Linux 16.04): macOS Monterey 12.1
- [ ] OS architecture (e.g. amd64): AMD64
- [ ] pyenv version: 2.2.3
- [ ] Python version: system default 3.8.9
- [ ] C Compiler information (e.g. gcc 7.3): gcc 4.2.1 / Apple clang 13.0.0 (clang-1300.0.29.30)
- [ ] Please attach the debug trace of the failing command as a gist:
 nothing to attach really... The dump doesn't output the error, but independent from the dump it causes the error.

```
eg@Eriks-MacBook-Pro ~ % env PYENV_DEBUG=1 which pyenv 2>&1 | tee trace.log
/usr/local/bin/pyenv
eg@Eriks-MacBook-Pro ~ % which pyenv
pyenv () {
	local command
	command="${1:-}" 
	if [ "$#" -gt 0 ]
	then
		shift
	fi
	case "$command" in
		(rehash | shell) eval "$(pyenv "sh-$command" "$@")" ;;
		(*) command pyenv "$command" "$@" ;;
	esac
}
```

**Comment by dhruvSHA256** [comment]
TLDR: no it does't break

From `man which` :
`which - shows the full path of (shell) commands.`

The code snippet you provided makes a function named pyenv which delegates the loading (aka lazy loading) of "actual pyenv" binary untill it is called, reducing the startup time of zsh. To check that call pyenv in a new shell, this will now load the actual pyenv, then run `which pyenv`. That will output the path of pyenv.

**Comment by alwayserik** [comment]
'which pyenv' works as you say it should... It works with zsh and bash shells.

What I assumed to be the problem is that 'which pyenv' doesn't enumerate the path when pyenv is setup in .zprofile  'eval "$(pyenv init --path)"'. I guess this is does what you call lazy loading? But if it lazy loads upon call, why wouldn't it do all the proper loading when we call it via 'which'?

**Comment by dhruvSHA256** [comment]
>  why wouldn't it do so when we call it via which
which just returns the path or definition of pyenv by checking it in your $PATH variable, it doesn't execute the command

**Comment by condekind** [comment]
> why wouldn't it do so when we call it via `which`?

Because there's no eval $(...) around the which. If you're using zsh, you should prefer `whence` anyway (and even if you were on bash, [one should avoid which](https://mywiki.wooledge.org/BashFAQ/081)). If you want to force a path search:

```sh
# zsh
whence -p pyenv  # Obs.: the p is lowercase here

# bash
type -P pyenv    # Obs.: the P is uppercase here
```

Just remember: if pyenv (or any other command) is a shell function, the program path returned by the commands above won't be executed. If you want to know the order considered by zsh for any given command:

```sh
# zsh
whence -fav myCommand

# bash
type -a myCommand
```

This issue should be closed, there's nothing broken.
